tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,317 @@
+    assert len(dataset) == num_samples, f"Got dataset length={len(dataset)} samples, expected {num_samples}"
+
+
+@pytest.mark.timeout(10)
+@pytest.mark.parametrize("created_ago", [0.5, 3])
+@pytest.mark.parametrize("timeout", [1])
+def test_reader_after_crash(tmpdir: py.path.local, created_ago: float, timeout: float):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = str(tmpdir.mkdir("remote"))
+    local = str(tmpdir.mkdir("local"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+
+    shutil.copy(os.path.join(remote, "index.mds"), os.path.join(local, "index.mds.tmp"))
+    shutil.copy(os.path.join(remote, "000003.mds"), os.path.join(local, "000003.mds.tmp"))
+    time.sleep(created_ago)
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders, timeout=timeout)
+
+    # Iterate over dataset and make sure there are no TimeoutErrors
+    for _ in dataset:
+        pass
+
+
+@pytest.mark.parametrize(
+    "share_remote_local",
+    [
+        True,
+        pytest.param(False, marks=pytest.mark.xfail(reason="__getitem__ currently expects shards to exist")),
+    ],
+)
+def test_reader_getitem(tmpdir: py.path.local, share_remote_local: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+
+    remote = str(tmpdir.mkdir("remote"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = str(tmpdir.mkdir("local"))
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders)
+
+    # Test retrieving random sample
+    try:
+        _ = dataset[17]
+    except Exception as e:
+        assert False, f"Unable to get random sample, got exception: {e}"
+
+
+@pytest.mark.daily()
+@pytest.mark.timeout(10)
+@pytest.mark.parametrize("batch_size", [1, 2, 5])
+@pytest.mark.parametrize("drop_last", [False, True])
+@pytest.mark.parametrize("num_workers", [1, 2, 3])
+@pytest.mark.parametrize("persistent_workers", [
+    False,
+    pytest.param(
+        True,
+        marks=pytest.mark.xfail(
+            reason=
+            "PyTorch DataLoader has non-deterministic worker cycle iterator when `persistent_workers=True`. Fixed in Mar 2022, likely landing PyTorch 1.12: https://github.com/pytorch/pytorch/pull/73675"
+        )),
+])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_dataloader_single_device(tmpdir: py.path.local, batch_size: int, drop_last: bool, num_workers: int,
+                                  persistent_workers: bool, shuffle: bool):
+    num_samples = 31
+    shard_size_limit = 1 << 6
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = str(tmpdir.mkdir("remote"))
```

Comment:
Can some of this code be refactored into fixtures? E.g. a dataset fixture? Seems to be similar to the test below.

BTW, this test is awesome -- gives me confidence in what is happening.